File: java/src/main/java/com/java/stack/ArrayStack.java
package com.java.stack;

public class ArrayStack<E> implements StackInterface<E> {
	private E stack[];
	private int topIndex;
	private static final int DEFAULT_CAPACITY = 64;
	private final E ERROR = null;


	//생성자1
	public ArrayStack() {
		stack = (E[]) new Object[DEFAULT_CAPACITY];
		topIndex = -1;
	}

	//생성자2
	public ArrayStack(int n) {
		stack = (E[]) new Object[n];
		topIndex = -1;
	}


	public static void main(String[] args) {
		ArrayStack<Integer> s = new ArrayStack<>(10);

		s.push(300);
		s.push(400);
		s.push(500);

		System.out.println(s.pop());
		System.out.println(s.top());

		System.out.println(s.isEmpty());
		s.popAll();

		System.out.println(s.isEmpty());
	}


	@Override
	public void push(E newItem) {
		if (isFull()) {
			System.out.println("stack over flow");
		} else {
			stack[++topIndex] = newItem;
		}

	}

	@Override
	public E pop() {
		if (isEmpty()) {
			System.out.println("stack is empty");
		} else {
			return stack[topIndex--];
		}
		return null;
	}

	@Override
	public E top() {
		if (isEmpty()) {
			System.out.println("stack is empty");
		} else {
			return stack[topIndex];
		}
		return null;
	}


	@Override
	public boolean isEmpty() {
		return topIndex < 0;
	}


	public boolean isFull() {
		return topIndex >= stack.length-1 ;
	}

	@Override
	public void popAll() {
		topIndex = -1;
		stack = (E[]) new Object[stack.length];
	}
}

File: kotlin/src/main/kotlin/com/kotlin/stack/ArrayStack.kt
package com.kotlin.stack

class ArrayStack (
    sizeOfArray: Int = 6
) : StackInterface {
    private var stack: Array<Int?>
    private var topIndex: Int
    private val error = -12345
    private var sizeOfArray = 6

    // 생성자
    init {
        stack = arrayOfNulls(sizeOfArray)
        topIndex = -1
        this.sizeOfArray = sizeOfArray
    }


    // push : 배열 마지막에 아이템 추가
    override fun push(newItem: Int) {
        if (isFull()) {
            println("array is full")
        } else {
            stack[++topIndex] = newItem
        }

    }

    // pop : 배열 마지막 아이템 return 후 삭제
    override fun pop(): Int? {
        if (isEmpty()) {
            println("array is empty")
        } else {
            return stack[topIndex--]
        }
        return error
    }

    // top : 배열 마지막 아이템 return만
    override fun top(): Int? {
        if (isEmpty()) {
            println("array is empty")
        } else {
            return stack[topIndex]
        }
        return error
    }

    // isEmpty
    override fun isEmpty(): Boolean {
        return topIndex < 0
    }

    // isFull
    private fun isFull(): Boolean {
        return topIndex >= stack.size-1
    }

    // popAll
    override fun popAll() {
        topIndex = -1
        stack = arrayOfNulls(stack.size)
    }




}